crates/uniswap-v4/src/uniswap/pool_providers/canonical_state_adapter.rs

```diff
@@ -28,23 +29,58 @@ impl CanonicalStateAdapter {
 }
 
 impl PoolManagerProvider for CanonicalStateAdapter {
-    fn subscribe_blocks(&self) -> futures::stream::BoxStream<Option<u64>> {
+    fn subscribe_blocks(&self) -> futures::stream::BoxStream<Option<PoolMangerBlocks>> {
         futures_util::stream::unfold(
             self.canon_state_notifications.resubscribe(),
             move |mut notifications| async move {
                 if let Ok(notification) = notifications.recv().await {
                     let mut last_log_write = self.last_logs.write().await;
                     let block = match notification {
-                        CanonStateNotification::Commit { new }
-                        | CanonStateNotification::Reorg { new, .. } => {
+                        CanonStateNotification::Commit { new } => {
                             let block = new.tip();
                             let logs: Vec<Log> = new
                                 .execution_outcome()
                                 .logs(block.number)
                                 .map_or_else(Vec::new, |logs| logs.cloned().collect());
                             *last_log_write = logs;
                             self.last_block_number.store(block.number, Ordering::SeqCst);
-                            Some(Some(block.number))
+                            Some(Some(PoolMangerBlocks::NewBlock(block.block.number)))
+                        }
+                        CanonStateNotification::Reorg { old, new } => {
+                            let tip = new.tip().block.number;
+                            // search 30 blocks back;
+                            let start = tip - 30;
+
+                            let mut range = old
+                                .blocks_iter()
+                                .filter(|b| b.block.number >= start)
+                                .zip(new.blocks_iter().filter(|b| b.block.number >= start))
+                                .filter(|&(old, new)| (old.block.hash() != new.block.hash()))
+                                .map(|(_, new)| new.block.number)
+                                .collect::<Vec<_>>();
+
+                            let range = match range.len() {
+                                0 => None,
+                                1 => {
+                                    let r = range.remove(0);
+                                    Some(r..=r)
+                                }
+                                _ => {
```

Comment:
Silly micro-optimization but you can use the `first()` and `last()` mechanism for a length of 1 as well

Reply:
ah true